File: src/services/githubService.ts
/**
 * GitHub API service layer.
 * All GitHub API calls go through this module.
 */

export interface GitHubRepositoryFile {
  path: string;
  sha: string;
  content: string;
}

export class GitHubApiError extends Error {
  constructor(
    public statusCode: number,
    message: string,
    public rateLimited: boolean = false,
  ) {
    super(message);
    this.name = "GitHubApiError";
  }
}

// ─── Helpers ──────────────────────────────────────────────────────────────────

function getToken(): string {
  const token = process.env.GITHUB_TOKEN;
  if (!token) {
    throw new GitHubApiError(500, "GITHUB_TOKEN is not configured");
  }
  return token;
}

/**
 * Parses "owner/repo" from a repoName or a full GitHub URL.
 */
export function parseOwnerRepo(input: string): { owner: string; repo: string } {
  // Handle full URL: https://github.com/owner/repo
  const urlMatch = input.match(/github\.com\/([^/]+)\/([^/.]+)/);
  if (urlMatch) {
    return { owner: urlMatch[1], repo: urlMatch[2] };
  }
  // Handle "owner/repo" format
  const parts = input.split("/");
  if (parts.length === 2 && parts[0] && parts[1]) {
    return { owner: parts[0], repo: parts[1] };
  }
  throw new GitHubApiError(400, `Cannot parse owner/repo from: ${input}`);
}

async function githubRest(
  path: string,
  options: RequestInit = {},
): Promise<Response> {
  const token = getToken();
  const res = await fetch(`https://api.github.com${path}`, {
    ...options,
    headers: {
      Authorization: `Bearer ${token}`,
      Accept: "application/vnd.github+json",
      "X-GitHub-Api-Version": "2022-11-28",
      ...options.headers,
    },
  });

  if (res.status === 403 && res.headers.get("x-ratelimit-remaining") === "0") {
    throw new GitHubApiError(429, "GitHub API rate limit exceeded", true);
  }
  if (res.status === 401) {
    throw new GitHubApiError(401, "GitHub token is invalid or expired");
  }
  if (res.status === 404) {
    throw new GitHubApiError(404, "Repository not found or not accessible");
  }

  return res;
}

function encodeGithubPath(path: string) {
  return path
    .split("/")
    .filter(Boolean)
    .map((part) => encodeURIComponent(part))
    .join("/");
}

export async function listRepositoryFiles(
  owner: string,
  repo: string,
  branch: string,
  basePath = "",
): Promise<Array<{ path: string; sha: string }>> {
  const normalizedBasePath = basePath.replace(/^\/+|\/+$/g, "");
  const ref = encodeURIComponent(branch);
  const res = await githubRest(`/repos/${owner}/${repo}/git/trees/${ref}?recursive=1`);

  if (!res.ok) {
    throw new GitHubApiError(res.status, `Failed to list repository files: ${res.statusText}`);
  }

  const tree = await res.json() as {
    tree?: Array<{ path: string; type: string; sha: string }>;
  };

  return (tree.tree ?? []).filter((entry) => {
    if (entry.type !== "blob") {
      return false;
    }

    if (!normalizedBasePath) {
      return true;
    }

    return entry.path === normalizedBasePath || entry.path.startsWith(`${normalizedBasePath}/`);
  }).map((entry) => ({
    path: entry.path,
    sha: entry.sha,
  }));
}

export async function getRepositoryFile(
  owner: string,
  repo: string,
  branch: string,
  path: string,
): Promise<GitHubRepositoryFile | null> {
  const encodedPath = encodeGithubPath(path);
  const ref = encodeURIComponent(branch);

  try {
    const res = await githubRest(`/repos/${owner}/${repo}/contents/${encodedPath}?ref=${ref}`);
    if (!res.ok) {
      throw new GitHubApiError(res.status, `Failed to fetch repository file: ${res.statusText}`);
    }

    const file = await res.json() as {
      path: string;
      sha: string;
      content?: string;
      encoding?: string;
    };

    const decodedContent =
      file.encoding === "base64" && file.content
        ? Buffer.from(file.content.replace(/\n/g, ""), "base64").toString("utf8")
        : "";

    return {
      path: file.path,
      sha: file.sha,
      content: decodedContent,
    };
  } catch (error) {
    if (error instanceof GitHubApiError && error.statusCode === 404) {
      return null;
    }

    throw error;
  }
}

export async function getRepositoryFileLastCommitDate(
  owner: string,
  repo: string,
  branch: string,
  path: string,
): Promise<string | null> {
  const encodedPath = encodeGithubPath(path);
  const ref = encodeURIComponent(branch);
  const res = await githubRest(
    `/repos/${owner}/${repo}/commits?path=${encodedPath}&sha=${ref}&per_page=1`,
  );

  if (!res.ok) {
    throw new GitHubApiError(
      res.status,
      `Failed to fetch repository file history: ${res.statusText}`,
    );
  }

  const commits = await res.json() as Array<{
    commit?: { author?: { date?: string | null } | null } | null;
  }>;

  return commits[0]?.commit?.author?.date ?? null;
}

export async function upsertRepositoryFile(
  owner: string,
  repo: string,
  branch: string,
  path: string,
  content: string,
  message: string,
) {
  const existing = await getRepositoryFile(owner, repo, branch, path);
  const encodedPath = encodeGithubPath(path);

  const res = await githubRest(`/repos/${owner}/${repo}/contents/${encodedPath}`, {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      message,
      content: Buffer.from(content, "utf8").toString("base64"),
      branch,
      ...(existing ? { sha: existing.sha } : {}),
    }),
  });

  if (!res.ok) {
    const body = await res.text();
    throw new GitHubApiError(res.status, `Failed to upsert repository file: ${body}`);
  }

  const result = await res.json() as {
    content?: { path: string; sha: string };
  };

  return {
    path: result.content?.path ?? path,
    sha: result.content?.sha ?? existing?.sha ?? "",
    created: !existing,
  };
}

export async function deleteRepositoryFile(
  owner: string,
  repo: string,
  branch: string,
  path: string,
  message: string,
) {
  const existing = await getRepositoryFile(owner, repo, branch, path);
  if (!existing) {
    return;
  }

  const encodedPath = encodeGithubPath(path);
  const res = await githubRest(`/repos/${owner}/${repo}/contents/${encodedPath}`, {
    method: "DELETE",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      message,
      branch,
      sha: existing.sha,
    }),
  });

  if (!res.ok) {
    const body = await res.text();
    throw new GitHubApiError(res.status, `Failed to delete repository file: ${body}`);
  }
}

File: backend/app/services/github_service.py
import httpx
import base64
import structlog
from app.config import get_settings

logger = structlog.get_logger()

GITHUB_API_BASE = "https://api.github.com"


class GitHubApiError(Exception):
    def __init__(self, message: str, status_code: int = 0, rate_limited: bool = False):
        self.message = message
        self.status_code = status_code
        self.rate_limited = rate_limited
        super().__init__(message)


def _get_headers() -> dict[str, str]:
    settings = get_settings()
    if not settings.GITHUB_TOKEN:
        raise GitHubApiError("GitHub token not configured", status_code=503)
    return {
        "Authorization": f"Bearer {settings.GITHUB_TOKEN}",
        "Accept": "application/vnd.github.v3+json",
        "X-GitHub-Api-Version": "2022-11-28",
    }


def parse_owner_repo(input_str: str) -> tuple[str, str]:
    input_str = input_str.strip().rstrip("/")
    if "github.com" in input_str:
        parts = input_str.split("github.com/")[-1].split("/")
        if len(parts) >= 2:
            return parts[0], parts[1]
    parts = input_str.split("/")
    if len(parts) == 2:
        return parts[0], parts[1]
    raise GitHubApiError(f"Cannot parse owner/repo from: {input_str}")


async def list_repository_files(
    owner: str, repo: str, branch: str = "main", base_path: str = ""
) -> list[dict]:
    url = f"{GITHUB_API_BASE}/repos/{owner}/{repo}/git/trees/{branch}?recursive=1"
    async with httpx.AsyncClient(timeout=30.0) as client:
        response = await client.get(url, headers=_get_headers())
        if response.status_code == 429:
            raise GitHubApiError("Rate limited", status_code=429, rate_limited=True)
        response.raise_for_status()
        data = response.json()
        tree = data.get("tree", [])
        if base_path:
            tree = [f for f in tree if f["path"].startswith(base_path)]
        return [{"path": f["path"], "type": f["type"], "size": f.get("size", 0)} for f in tree]


async def get_repository_file(
    owner: str, repo: str, branch: str, path: str
) -> str:
    url = f"{GITHUB_API_BASE}/repos/{owner}/{repo}/contents/{path}?ref={branch}"
    async with httpx.AsyncClient(timeout=30.0) as client:
        response = await client.get(url, headers=_get_headers())
        if response.status_code == 404:
            raise GitHubApiError(f"File not found: {path}", status_code=404)
        if response.status_code == 429:
            raise GitHubApiError("Rate limited", status_code=429, rate_limited=True)
        response.raise_for_status()
        data = response.json()
        content = data.get("content", "")
        return base64.b64decode(content).decode("utf-8")


async def get_file_last_commit_date(
    owner: str, repo: str, branch: str, path: str
) -> str | None:
    url = f"{GITHUB_API_BASE}/repos/{owner}/{repo}/commits?sha={branch}&path={path}&per_page=1"
    async with httpx.AsyncClient(timeout=30.0) as client:
        response = await client.get(url, headers=_get_headers())
        response.raise_for_status()
        commits = response.json()
        if commits:
            return commits[0]["commit"]["committer"]["date"]
        return None


async def upsert_repository_file(
    owner: str, repo: str, branch: str, path: str, content: str, message: str
) -> dict:
    url = f"{GITHUB_API_BASE}/repos/{owner}/{repo}/contents/{path}"
    headers = _get_headers()

    sha = None
    async with httpx.AsyncClient(timeout=30.0) as client:
        check = await client.get(f"{url}?ref={branch}", headers=headers)
        if check.status_code == 200:
            sha = check.json().get("sha")

        payload = {
            "message": message,
            "content": base64.b64encode(content.encode("utf-8")).decode("ascii"),
            "branch": branch,
        }
        if sha:
            payload["sha"] = sha

        response = await client.put(url, json=payload, headers=headers)
        response.raise_for_status()
        return response.json()


async def delete_repository_file(
    owner: str, repo: str, branch: str, path: str, message: str
) -> dict:
    url = f"{GITHUB_API_BASE}/repos/{owner}/{repo}/contents/{path}"
    headers = _get_headers()

    async with httpx.AsyncClient(timeout=30.0) as client:
        check = await client.get(f"{url}?ref={branch}", headers=headers)
        if check.status_code == 404:
            raise GitHubApiError(f"File not found: {path}", status_code=404)
        check.raise_for_status()
        sha = check.json()["sha"]

        payload = {"message": message, "sha": sha, "branch": branch}
        response = await client.delete(url, json=payload, headers=headers)
        response.raise_for_status()
        return response.json()
